bundles/org.eclipse.swt/Eclipse SWT Custom Widgets/common/org/eclipse/swt/custom/BusyIndicator.java

```diff
@@ -70,6 +70,41 @@ public static void showWhile(Display display, Runnable runnable) {
 		}
 	}
 
+	/**
+	 * If called from a {@link Display} thread, waits for the given
+	 * <code>Future</code> to complete and provides busy feedback using the
+	 * busy indicator. While waiting for completion, pending UI events are processed
+	 * to prevent UI freeze.
+	 *
+	 * If there is no {@link Display} for the current thread, the method returns
+	 * immediately and no feedback is shown to the user.
+	 *
+	 * In any case, the passed {@link Future} is returned for chaining of further operations.
+	 *
+	 * @param future the {@link Future} for which busy feedback is to be shown.
+	 * @return the {@link Future} for chaining.
+	 * @since 3.127
+	 */
+	public static <F extends Future<T>, T> F showWhile(F future) {
```

Comment:
I'm afraid that by returning the future, the expectation would be that this method returns immediately (as it's a usual pattern), whatever documentation says.

Reply:
If one wants that, there are two other methods that exactly do that.

Reply:
I‘d return T instead of F to clearly communicate that this is blocking

Reply:
Future chaining based on a already done future appears to be contrived usecase

Reply:
You can do so much things on a completed future... also please keep in mind that *getting* a future value has different effects (e.g. blocking on an arbitrary thread) compared to just checking it is done (what can be implemented lock-free).

Reply:
IMO, it'd be much safer to drop the chaining, which can be counter-intuitive here and thus error-prone, and just make the method void.

Reply:
Chaining is very common with `CompletableFutures` so I don't see how/why it should be error prone even though it is not completely required it is very convenient.

Reply:
Yes chaining is convenient, but in the common approach of CompletableFuture, the composition of chaining and waiting with the same method seems very unusual and not easy to understand. I'm not against chaining, nor against waiting, but both at the same time is error-prone as it's not what the existing APIs of CompletableFuture has build as habit.

Reply:
>  I'm not against chaining, nor against waiting, but both at the same time is error-prone as it's not what the existing APIs of CompletableFuture has build as habit.

I couldn't have voiced my concerns better.
The proposed API is against all "trained" behavior when it comes to APIs that consume or return futures.

Reply:
To make decision easier I would like to ask reviewers to give a "thumb up" for what they prefer:

1.  [here](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1266#discussion_r1642496049) for keeping "chaining" and return the future as is
2. [here](https://github.com/eclipse-platform/eclipse.platform.swt/pull/1266#discussion_r1642509046) for changing method to void 

I'll then adjust simply to the majority vote then, I don't want to consider "return `T`" as in my experience this often is hard to code "right" e.g in this case of "non ui thread" I would need to block what i'd like to avoid.

Reply:
> The proposed API is against all "trained" behavior when it comes to APIs that consume or return futures.

Could you give  examples? e.g. `CompletableFuture.completedFuture(`...) and alike obviously return an already completed future, so I can't tell any other "API" that consumes things so I don't think there is any real "specified behavior" of functions beside "its useful"...

Reply:
@laeubi we can take the input, but here I think there is a clear hint of one being quite error-prone vs one being safe. Running a popularity contest is not known to be the right approach to deal with risk mitigation and technical debt management.

Reply:
> CompletableFuture.completedFuture(...) and alike 

Can you share an example of what the `alike` are? This method seems to me as the very one and very explicit exception in the whole CompletableFuture design, and this method is not the most usual one consumers would use. There are ~30 other methods in that interface that return the future immediately and which are shown in every tutorial about CompletableFuture.

Reply:
I don't se the risk (what risk?) / error prone (what could go wrong?), the effect is simply the same (method will show busy if in UI thread) , the only difference is now:

```
BusyIndicator.showWhile(future).thenRunAsync(() -> {
    if (text.isDisposed() || b2.isDisposed()) {
	    return;
    }
    b2.setEnabled(true);
    text.append("\nBackground work has completed");
}, display);
```


versus

```
BusyIndicator.showWhile(future);
future.thenRunAsync(() -> {
    if (text.isDisposed() || b2.isDisposed()) {
	    return;
    }
    b2.setEnabled(true);
    text.append("\nBackground work has completed");
}, display);
```

or equivalently:

```
BusyIndicator.showWhile(future.thenRunAsync(() -> {
    if (text.isDisposed() || b2.isDisposed()) {
	    return;
    }
    b2.setEnabled(true);
    text.append("\nBackground work has completed");
}, display));
```

I really can't see anything is "error prone" or risky before/after the change to void

Reply:
@mickaelistria I think we simply prefer different style, I can accept that and don't think its worth to block the further processing here so I would just change the implementation if no one agrees, its really only a matter of chained calls versus consecutive lines ...

Reply:
The error comes from the fact that the extreme majority of developers expect `Future<F> blah(Future<F>)` to return immediately because it's how CompletableFuture API was designed, and allowing to return immediately without waiting is the main reason why CompletableFuture were created.

Reply:
Exactly. What's wrong with meeting the expectation and changing the implementation accordingly?

```java
public static <T> CompletableFuture<T> showWhile(CompletableFuture<T> future) {
  Display display = ...
  Integer busyId = setBusyCursor(display);
  // avoid the busy loop from within the composeAsync
  future.whenComplete((any, anyex)->display.wake()); 
  return CompletableFuture.completedFuture(busyId).thenComposeAsync((id)->{
    try {
      while (!future.isDone() && !display.isDisposed()) {
        if (!display.readAndDispatch()) {
          Display.sleep();
        }
      }
     return future; // return the future itself to avoid the hassle with the exception handling here
    } finally {
      clearBusyCursor(display, id);
    }
  }, display);
}



```

Reply:
Well that's possibly but completely strange and complex way to rewrite what the method currently does (except some variations) ;-)

For that just let us follow the flow:

1. we are on the UI thread
2. we have a future running in a different thread
3. we call `showWhile(...)`
4. we set the busy cursor
5. the compose is executed with the current display as executor, that will result in a call to display.syncExec (because we are on the UI thread) what the will block until the future is done (+sleep for undefined time if no UI event occurs)
6. then we will return the future
7. the we will wake the UI thread (that is the eventqueue of the original caller)

okay we might "heal" that by using Display:asyncExec explicitly, but for what purpose? we will return the future but nothing will happen unless the caller leaves the UI thread, the async processing takes place and takes over and the caller *must* use the future to enter the event queue again.

So what we effectively do is to do the same, but we require it another round trip through the event queue, just for the sake of returning a future that the caller already has at hand before.... 

So the whole point of this is, we are already in the UI thread and we want to wait but do not make UI unresponsive, yes that's very special, but that's how SWT was designed (or maybe many UI frameworks) but because of this you can't apply "usual" patterns... at least not if you don't like deadlocks or UI freeze ;-)

Reply:
Ok, I back out of the discussion. You're completely neglecting the argument regarding the expectation for such an API (that it returns immediately) and it seems you only argue that the implementation of this utility is simpler though this method is only written once yet all the users must keep in mind that this pipelineable future method is blocking the caller thread until the underlying future is done compared to the expectation that pipelining futures is usually assembling a future that will eventually be awaited. I lack to see a basis for a discussion here.

Reply:
> completely strange 

one might argue the same about a method that will loop until a future is done for the sake of returning the future afterwards. But who am I to judge.

Reply:
I now completely aligned the method with the "traditional" `BusyIndicator.showWhile(Display, Runnable)` so it blocks always and not returns anything.

Reply:
> You're completely neglecting the argument regarding the expectation for such an API (that it returns immediately) and it seems you only argue that the implementation of this utility is simpler though this method is only written once yet all the users must keep in mind that this pipelineable future method is blocking the caller thread until the underlying future is done compared to the expectation that pipelining futures is usually assembling a future that will eventually be awaited. I lack to see a basis for a discussion here.

The problem is, this is (due to its nature) not a "pipeline-able method" ... I do not expsect everyone is using this methods, and they are badly documented (mostly by code) but 

- `showWhile` is and ever was blocking
- showWhile (+ `Feature` **(!)**) I initially designed to be "not always blocking", (now changed it is as "it was for 20 years" --> always blocking as well).
- `execute` / `compute` where designed non blocking and allows "usual" Completable feature semantics since day 1 already.